### ISSUE TITLE
argument for detect_anomaly on for debug

### ISSUE THREAD

**Issue by ruotianluo** [PyTorchLightning/pytorch-lightning#8313]
## 🚀 Feature
Add an arg for trainer to turn on detect_anomaly. So that when gradients get nan, the model will automatically detect.

### Motivation

For easier debugging.

**Comment by ananthsub** [comment]
@ruotianluo - there is `terminate_on_nan` on the Trainer constructor: https://github.com/PyTorchLightning/pytorch-lightning/blob/4928dc55790eec4ee4a558f494a6fc3d49a5e78a/pytorch_lightning/trainer/trainer.py#L304-L305

**Comment by ruotianluo** [comment]
It is actually different. detect anomaly tells you where you got the nan during back propagation instead of just terminate.

**Comment by ananthsub** [comment]
ah do you mean support for PyTorch's `torch.autograd.detect_anomaly` detection here? https://pytorch.org/docs/stable/_modules/torch/autograd/anomaly_mode.html#detect_anomaly

**Comment by ruotianluo** [comment]
yes

**Comment by ananthsub** [comment]
@awaelchli @carmocca what do you think? we could enable `torch.autograd.set_detect_anomaly(trainer.terminate_on_nan)` inside of the training loop.

**Comment by awaelchli** [comment]
Yes, nice Idea. @ruotianluo would you like to give this a try?

**Comment by ananthsub** [comment]
@awaelchli - one concern is the slowdown that will occur when using `terminate_on_nan` with and without anomaly detection. my assumption is that people use `terminate_on_nan` for debugging, but i'm not sure if that's true in all cases.

**Comment by awaelchli** [comment]
Yes, I would recommend strictly for debugging. I thought it was already there but we should add it to our debugging section in the docs. 

Perhaps `terminate_on_nan` should be replaced by anomaly detection entirely and the trainer argument renamed? According to the [pytorch docs](https://pytorch.org/docs/stable/autograd.html), their anomaly detection already terminates with an error. The difference is that PL stops also when the loss is NaN.

**Comment by carmocca** [comment]
Agree that this would be a great addition.

Does `detect_anomaly` also check for +-inf? The docs just say

> Any backward computation that generate “nan” value will raise an error.

**Comment by ananthsub** [comment]
fyi @daniellepintz - this would be another good issue & feature as we're looking at error handling & debugging capabilities

**Comment by daniellepintz** [comment]
sg! could someone to assign to me? unless @ruotianluo wants to take it on
